src/cclib/io/wfxwriter.py

```diff
@@ -0,0 +1,488 @@
+    closing_tag = ['</' + section_name + '>']
+
+    if isinstance(section_data, list):
+        return opening_tag + section_data + closing_tag
+    elif isinstance(section_data, str):
+        return opening_tag + (' ' + section_data).split('\n') + closing_tag
+    elif isinstance(section_data, int) or isinstance(section_data, float):
+        return opening_tag + [' ' + str(section_data)] + closing_tag
+
+def _list_format(data, per_line, style='%22.11E'):
+    """Format lists for pretty print."""
+    template = style * per_line
+    leftover = len(data) % per_line
+    # Template for last line.
+    last_template = style * leftover
+
+    pretty_list = [template % tuple(data[i:i+per_line])
+                    for i in range(0, len(data) - leftover, per_line)]
+    if leftover:
+        return pretty_list + [last_template % tuple(data[-1*leftover:])]
+    return  pretty_list
+
+class WFXWriter(filewriter.Writer):
+    """A writer for wfx files."""
+
+    required_attrs = ('atomcoords', 'atomnos', 'gbasis', 'charge', 'homos',
+                      'mult', 'mocoeffs')
+    keyword = 'GTO'
+
+    def _title(self):
+        """Return filename without extension to be used as title."""
+        title = "Written by cclib-1.5."
+        if self.jobfilename is not None:
+            return os.path.basename(os.path.splitext(self.jobfilename)[0]) +\
+                    '. ' + title
+        return title
+
+    def _keywords(self):
+        """Return one of GTO, GIAO, CSGT keyword."""
+        # AIMALL currently supports on GTO keyword.
+        return self.keyword
+
+    def _no_of_nuclei(self):
+        """Number of nuclei in the molecule."""
+        return len(self.ccdata.atomnos)
+
+    def _no_of_prims(self):
+        """Number of Primitives."""
+        nprims = 0
+        for atom in self.ccdata.gbasis:
+            for prims in atom:
+                nprims += num_orb[prims[0]] * len(prims[1])
+        return nprims
+
+    def _no_of_mos(self):
+        """Number of occupied MOs."""
+        return int(max(self.ccdata.homos)) + 1
+
+    def _no_of_perturbations(self):
+        """Number of Perturbation.
+
+        This is usually zero.  For GIAO it should be 3
+        (corresponding to Lx, Ly and Lz), and
+        for CSGT it should be 6
+        (corresponding to Lx, Ly, Lz, Px, Py and Pz).
+        """
+        if 'GIAO' in self._keywords():
+            return 3
+        elif 'CSGT' in self._keywords():
+            return 6
+        return 0
+
+    def _nuclear_names(self):
+        """Names of nuclei present in the molecule.
+
+        O1
+        H2
+        H3
+        """
+        return [self.pt.element[Z]+str(i) for i, Z in
+                enumerate(self.ccdata.atomnos, start=1)]
+
+    def _atomic_nos(self):
+        """Atomic numbers of elements."""
+        return [str(Z) for Z in self.ccdata.atomnos]
+
+    def _nuclear_charges(self):
+        """Nuclear charges."""
+        if hasattr(self.ccdata, 'coreelectrons'):
```

Comment:
Please calculate the charges and then return at end.

Reply:
Refactored to now have a single return.